gpbft/gpbft.go

```diff
@@ -1374,29 +1374,54 @@ func newBroadcastState() *broadcastState {
 // record stores messages that are required should rebroadcast becomes necessary.
 // The messages stored depend on the current progress of instance. If the
 // instance progresses to DECIDE then only the decide message will be recorded.
-// Otherwise, all messages except QUALITY from the current and previous rounds
+// Otherwise, all messages including QUALITY from the current and previous rounds
 // (i.e. the latest two rounds) are recorded.
 //
 // Note, the messages recorded are more than what FIL-00896 strictly requires at
 // the benefit of less reliance on rebroadcast and reduction in the possibility
 // of participants getting stuck.
+//
+// The rationale for including QUALITY messages as part of rebroadcast is to
+// improve censorship resistance at the face of unreliable message delivery,
+// which can ultimately result in lack of instance progress for extended periods
+// of time. See:
+//   - https://github.com/filecoin-project/FIPs/discussions/809#discussioncomment-10409902
+//   - https://github.com/filecoin-project/FIPs/discussions/809#discussioncomment-10424988
 func (bs *broadcastState) record(mb *MessageBuilder) {
 	switch mb.Payload.Step {
 	case DECIDE_PHASE:
 		// Clear all previous messages, as only DECIDE message need to be rebroadcasted.
 		// Note that DECIDE message is not associated to any round, and is always
 		// broadcasted using round zero.
-		bs.messagesByRound = make(map[uint64][]*MessageBuilder)
+		clear(bs.messagesByRound)
 		bs.messagesByRound[0] = []*MessageBuilder{mb}
-	case COMMIT_PHASE, PREPARE_PHASE, CONVERGE_PHASE:
+	case QUALITY_PHASE, CONVERGE_PHASE, PREPARE_PHASE, COMMIT_PHASE:
```

Comment:
would it be simpler to handle quality messages separately? Really, decide messages as well? Neither belong to any given round.

Reply:
I had a bunch of discussions in previous reviews on this. Originally I did implement it using dedicated fields. But it ended up being a map as that seemed simpler. 

I plan to change it back to that approach as part of making rebroadcasted messages more selective.

I am happy to reflect on your comment in this PR if it's a blocker. Otherwise I leave it to future PRs if the code here is correct. 

What is your preference?

Reply:
It's not, I'll review as-is.